arbor/morph/region.cpp

```diff
@@ -0,0 +1,306 @@
+    }
+    return region(cable_{c});
+}
+
+region branch(msize_t bid) {
+    return cable({bid, 0, 1});
+}
+
+mcable_list thingify_(const cable_& reg, const em_morphology& em) {
+    auto& m = em.morph();
+
+    if (reg.cable.branch>=m.num_branches()) {
+        throw morphology_error(util::pprintf("Branch {} does not exist in morpology", reg.cable.branch));
+    }
+
+    return {reg.cable};
+}
+
+std::ostream& operator<<(std::ostream& o, const cable_& c) {
+    return o << c.cable;
+}
+
+//
+// region with all segments with the same numeric tag
+//
+struct tagged_ {
+    int tag;
+};
+
+region tagged(int id) {
+    return region(tagged_{id});
+}
+
+mcable_list thingify_(const tagged_& reg, const em_morphology& em) {
+    auto& m = em.morph();
+    size_t nb = m.num_branches();
+
+    std::vector<mcable> L;
+    L.reserve(nb);
+    auto& samples = m.samples();
+    auto matches     = [reg, &samples](msize_t i) {return samples[i].tag==reg.tag;};
+    auto not_matches = [&matches](msize_t i) {return !matches(i);};
+
+    for (msize_t i: util::make_span(nb)) {
+        auto ids = util::make_range(m.branch_indexes(i)); // range of sample ids in branch.
+        size_t ns = util::size(ids);        // number of samples in branch.
+
+        if (ns==1) {
+            // The branch is a spherical soma
+            if (samples[0].tag==reg.tag) {
+                L.push_back({0,0,1});
+            }
+            continue;
+        }
+
+        // The branch has at least 2 samples.
+        // Start at begin+1 because a segment gets its tag from its distal sample.
+        auto beg = std::cbegin(ids);
+        auto end = std::cend(ids);
+
+        // Find the next sample that matches reg.tag.
+        auto start = std::find_if(beg+1, end, matches);
+        while (start!=end) {
+            // find the next sample that does not match reg.tag
+            auto first = start-1;
+            auto last = std::find_if(start, end, not_matches);
+
+            auto l = first==beg? 0.: em.sample2loc(*first).pos;
+            auto r = last==end?  1.: em.sample2loc(*(last-1)).pos;
+            L.push_back({i, l, r});
+
+            // Find the next sample in the branch that matches reg.tag.
+            start = std::find_if(last, end, matches);
+        }
+    }
+    return L;
+}
+
+std::ostream& operator<<(std::ostream& o, const tagged_& t) {
+    return o << "(tag " << t.tag << ")";
+}
+
+//
+// region with all segments in a cell
+//
+struct all_ {};
+
+region all() {
+    return region(all_{});
+}
+
+mcable_list thingify_(const all_&, const em_morphology& m) {
+    auto nb = m.morph().num_branches();
+    mcable_list branches;
+    branches.reserve(nb);
+    for (auto i: util::make_span(nb)) {
+        branches.push_back({i,0,1});
+    }
+    return branches;
+}
+
+std::ostream& operator<<(std::ostream& o, const all_& t) {
+    return o << "all";
+}
+
+//
+// intersection of two point sets
+//
```

Comment:
Copy-pasto on the comment here and on line 269: should be regions rather than point sets.